crates/metadata-store/src/lib.rs

```diff
@@ -164,9 +165,89 @@ impl MetadataStoreClient {
             .await
     }
 
-    async fn delete(&self, key: ByteString, precondition: Precondition) -> Result<(), WriteError> {
+    pub async fn delete(
+        &self,
+        key: ByteString,
+        precondition: Precondition,
+    ) -> Result<(), WriteError> {
         self.inner.delete(key, precondition).await
     }
+
+    pub async fn read_modify_write<T, F>(
+        &self,
+        key: ByteString,
+        mut modify: F,
+    ) -> Result<T, ReadModifyWriteError>
+    where
+        T: Versioned + Serialize + DeserializeOwned + Clone,
+        F: FnMut(Option<T>) -> Operation<T>,
+    {
+        loop {
+            let value = self.get::<T>(key.clone()).await?;
+
+            let precondition = value
+                .as_ref()
+                .map(|c| Precondition::MatchesVersion(c.version()))
+                .unwrap_or(Precondition::DoesNotExist);
+
+            let result = modify(value);
+
+            match result {
+                Operation::Upsert(value) => {
+                    match self.put(key.clone(), value.clone(), precondition).await {
+                        Ok(()) => return Ok(value),
+                        Err(WriteError::FailedPrecondition(msg)) => {
+                            debug!("concurrent value update: {msg}; retrying");
+                        }
+                        Err(err) => return Err(err.into()),
+                    }
+                }
+                Operation::Return(value) => return Ok(value),
+            }
+        }
+    }
+}
+
+pub enum Operation<T> {
+    /// Upsert the provided value and return if successful.
+    Upsert(T),
+    /// Return the provided value w/o upserting it.
+    Return(T),
+}
+
+#[derive(Debug, thiserror::Error)]
+pub enum ReadModifyWriteError {
+    #[error("network error: {0}")]
+    Network(GenericError),
+    #[error("internal error: {0}")]
+    Internal(String),
+    #[error("serialization failed: {0}")]
+    Serialization(GenericError),
+    #[error("deserialization failed: {0}")]
+    Deserialization(GenericError),
```

Comment:
Many parts of the code call this Codec, no strong opinions, but should we standardize on one?

Reply:
Let's standardize on `Codec` then. Consistency is king.